### DISCUSSION TITLE
Use Pydantic(?) to map returned API response field names to internal ORM names

### DISCUSSION THREAD

**Discussion by wakatara** [fastapi/fastapi#10402] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
type struct ephemeris {
     phase_angle:   float  `json: STO`
     true_anomoly:  float  `json: TRUEANOM`
     etc
     etc


This issue seemed *similar* but felt more about having people post *to* the application with different names rather than receiving data from *other* applications and mapping it to internal field names. Please let me know if I am mistaken in my assumption here.
For pretty obvious reasons, I'd like the received data to map to None or that particular field so as to remove boilerplate logic which just junks up things. 

https://github.com/tiangolo/fastapi/discussions/7742
```

### Description

I have several "foreign" APIs that return results for an astronomy project I am working on. 

Often these fields are porly-named/capitalized/foreshortened/acronym names of fields that *should* be named better (and furthermore in some cases heavily nested in json) I need to have in the database in better form. I am looking for a means with FastAPI, working with Pydantic (? since, it seems the lib to use for this) to map those fields via their json response to the schema names I would use in Pydantic (note: at the moment, I am using a schema.py and models.py file as I am using Pydantic _and_ SQLALchemy, rather than SQLModel on top of it all.).

I know in GoLang or Rust I would use serde or map the struct with json fields: 

```go
type struct ephemeris {
     phase_angle:   float  `json: STO`
     true_anomoly: float   `json: TRUEANOM`
     etc
     etc
```

to accomplish this. But at the moment I'm either misunderstanding how Python does this with Pydantic or have jsut become too used to struct based languages (and my google fu appeared not to being anything up on this, so I apologize if this is somehow in the docs somewhere and I was unable to surface this). 8-/. 

Is there a way to easily do this with FastAPI and Pydantic (or SQLModel) in some elegant fashion? At the moment, I'm needing to put a lot of parsing/missing keys/conversion in dict logic which seems clunky and I imagine this is perhaps a common use case for users of FastAPI. 

thanks! 
PS> Apologies if this is better asked in the Pydantic group, but it seemed much more appropriate here since directly related to how it's being used.

### Operating System

macOS

### Operating System Details

not relevant to issue. I do have FastAPI dockerized (which is the bomb btw. Really liking Fast API.).

### FastAPI Version

0.103.2

### Pydantic Version

1.10.13

### Python Version

3.11.6

**Comment by jgould22** [comment]
Pydantic is mostly concerned about validation so its not entirely analogous to something like serde though there is some overlap (underneath the python pydantic core is written in rust). 

I think what you are looking for is the Field Alias https://docs.pydantic.dev/latest/concepts/fields/#field-aliases and some of its associated settings (AliasPath, AliasChoices etc). It works sorta like the serde attributes.

Pydantic is not quite as explicit with separating serlization/deserialization as Serde so some options work both ways and it is a little less obvious.

**Comment by wakatara** [comment]
I did look at the Field stuff due to the other post I mentioned from my in-forum search, but how to apply it in the manner I've suggested, where it acts to broker mappings between inbound json and my canonical models still has me scratching my head. Have you ever seen any posts or xamples of such a thing? As mentioned, I had assumed this might be a common use case (and have to admit, I'm much more familiar with other languages and frameworks in Ruby, Go, and Rust than FastAPI (this is my first big project using FastAPi and sqlalchemy/pydantic in this fashion. 

Thanks for responding though! Appreciate it. If I do manage to crack it, I will write up a very detailed blog post so it adds to the corpus on FastAPI though.

**Reply by jgould22** [reply]
So to clarify

You have some external API that you are getting data from which returns the data in a json format.

The schema of this external API is not super intuitive and you would like you internal code to use different name/structure so it is easier to understand.

You would like to call this API and get a response.

Once you have the response you are wondering what is the easiest/most elegant way to go from a json response to your SQLAlechemy defined model.

Is that a correct interpretation ?

**Comment by wakatara** [comment]
Hey Jordon

Yes, you basically have it. Correct.

A little more than that, I'd like to almost have pydantic (or something else in python land) basically do the pattern matching declaratively. So, much as I would with sucking in a Go struct, with the added pydantic superpowers (ie. this field is not optional so if it's empty throw error or if there is no value here set the default to = None etc.). 

As Pydantic allows validation on the field types (the `Fields` feature I'm still grokking as to whether it can be bent ot what I want, or I'm just not getting it from the documentation and it *does* do this.). At the moment, the main issue is one api I get info back from in particular (as an example r refers to heliocentric_distance - only a pro would know r meant distance in au's etc. This the remap for explicitness.).

(and apologies for using a Go construct to demonstrate the idea I want. I imagine pythonistas have their own way of going about this, but I'm just having trouble figuring out what it might be.).

(the problem is not just one of preference but of the fact that some
